core/src/main/java/tc/oc/pgm/menu/InventoryMenu.java

```diff
@@ -1,151 +1,79 @@
 package tc.oc.pgm.menu;
 
-import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.base.Preconditions.checkNotNull;
-
-import java.util.WeakHashMap;
-import javax.annotation.Nullable;
-import org.bukkit.Bukkit;
-import org.bukkit.inventory.Inventory;
-import org.bukkit.inventory.ItemStack;
+import org.bukkit.event.inventory.ClickType;
 import tc.oc.pgm.api.player.MatchPlayer;
-import tc.oc.pgm.observers.ObserverToolsMatchModule;
-import tc.oc.pgm.util.StringUtils;
-import tc.oc.pgm.util.component.ComponentRenderers;
-import tc.oc.pgm.util.component.types.PersonalizedTranslatable;
-
-public abstract class InventoryMenu {
+import tc.oc.pgm.menu.items.InventoryItem;
 
-  protected static final int ROW_WIDTH = 9; // Number of columns per row
-  protected static final int MAX_ROWS = 6; // Max allowed row size
-
-  private final WeakHashMap<MatchPlayer, InventoryMenu> viewing =
-      new WeakHashMap<>(); // Map of players who are viewing the gui, along with the menu
-  private final String title; // Title of the inventory
-  private final int rows; // The # of rows in the inventory
+/** Interface that describes a programmable inventory menu. */
+public interface InventoryMenu {
```

Comment:
I like the interface, I think methods need a little tweaking. Right now there's 3 `open`s and a `purge` and `refresh`,  it's not clear what each of these do from a bird's eye view.

First, I'd say move this to `tc.oc.pgm.util.menu` and expand the scope from `MatchPlayer` to `Player`. There's no reason, that I can think of, that we should scope this to a `Match`. For instance, when we support multiple matches we might have some menu that allows you to switch between them. For settings, in the `onClick` callback, you can use `Datastore#getSettings(Player#getId)` to get the player's settings.

Next, I'd say you only need 1 open. Any cyclical back checks can be an implementation detail.
```java
/**
 * Opens the menu for a player.
 *
 * @param viewer A player.
 * @param referrer A previous menu that referred this menu, or null if the first menu.
 */
void open(Player viewer, @Nullable InventoryMenu referrer);
```

Finally, you really only need 1 type of invalidate, which basically just re-draws the items.
```java
/**
 * Refreshes the menu for a player, on next tick.
 *
 * @param viewer A player.
 */
void invalidate(Player viewer);
```

On the implementation side, you could use metadata for back checks or history. e.g. (a `Stack<InventoryMenu>` stored as metadata?)

Reply:
Doing history as an implementation detail makes it a bit messy. I'd rather build it into the API or not have it at all and I think for this case I'll just chop it out since, while relevant for Walrus, isn't necessary here. Everything else is good though, and in the future we can build a different implementation which adds pagination/history.